### PR TITLE
[Merged by Bors] - feat: port RingTheory.Finiteness

### PR DESCRIPTION
**Pull request by ChrisHughes24**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mattrobball** [comment]
Still has eta

**Comment by casavaca** [comment]
@Parcly-Taxel This should not be `awaiting-review` because it contains `set_option synthInstance.etaExperiment true in` which is for local testing only.

**Comment by jcommelin** [comment]
@casavaca that is not strictly true. If all other options have been tried, but we can't get the PR to compile without `etaExperiment` than we sometimes make an exception.

File: Mathlib/RingTheory/Finiteness.lean [hunk 1]
```diff
@@ -226,11 +226,14 @@ theorem fg_of_fg_map_injective (f : M →ₗ[R] P) (hf : Function.Injective f) {
       rw [← LinearMap.range_coe, ← span_le, ht, ← map_top]
       exact map_mono le_top⟩
 #align submodule.fg_of_fg_map_injective Submodule.fg_of_fg_map_injective
+/- Porting note: Lean not finding this instance is source of the majority of troubles here. 
+Perhaps it should get its own name -/
+alias LinearMap.instSemilinearMapClassLinearMap ← instSLMC
 
-set_option synthInstance.etaExperiment true in
 theorem fg_of_fg_map {R M P : Type _} [Ring R] [AddCommGroup M] [Module R M] [AddCommGroup P]
```
Comment (jcommelin) [inline Mathlib/RingTheory/Finiteness.lean:233]: With `etaExperiment`, we don't need the `@`s in the following statement.

File: Mathlib/RingTheory/Finiteness.lean [hunk 2]
```diff
@@ -265,25 +267,31 @@ theorem fg_pi {ι : Type _} {M : ι → Type _} [Finite ι] [∀ i, AddCommMonoi
...
+def asFun [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N] (f : M →ₗ[R] N) : M → N := f
+
+-- set_option synthInstance.etaExperiment true in
 /-- If 0 → M' → M → M'' → 0 is exact and M' and M'' are
 finitely generated then so is M. -/
 theorem fg_of_fg_map_of_fg_inf_ker {R M P : Type _} [Ring R] [AddCommGroup M] [Module R M]
-    [AddCommGroup P] [Module R P] (f : M →ₗ[R] P) {s : Submodule R M} (hs1 : (s.map f).Fg)
-    (hs2 : (s ⊓ LinearMap.ker f).Fg) : s.Fg := by
+    [AddCommGroup P] [Module R P] (f : M →ₗ[R] P) {s : Submodule R M} 
+    -- Porting note: Lean having trouble both unifying for SLMC and then synthesizing once unified
```
Comment (jcommelin) [inline Mathlib/RingTheory/Finiteness.lean:279]: With `etaExperiment`, we don't need the `@`s in the following statement.

File: Mathlib/RingTheory/Finiteness.lean [hunk 3]
```diff
@@ -265,25 +267,31 @@ theorem fg_pi {ι : Type _} {M : ι → Type _} [Finite ι] [∀ i, AddCommMonoi
...
+    (hs1 : (@map R R M P _ _ _ _ _ _ (RingHom.id _) _ _ instSLMC f s).Fg)
+    (hs2 : (s ⊓ @LinearMap.ker R R M P _ _ _ _ _ _ (RingHom.id _) _ instSLMC f).Fg) : s.Fg := by
   haveI := Classical.decEq R
   haveI := Classical.decEq M
   haveI := Classical.decEq P
   cases' hs1 with t1 ht1
   cases' hs2 with t2 ht2
-  have : ∀ y ∈ t1, ∃ x ∈ s, f x = y := by
+  have : ∀ y ∈ t1, ∃ x ∈ s, asFun f x = y := by
```
Comment (jcommelin) [inline Mathlib/RingTheory/Finiteness.lean:287]: With `etaExperiment`, we don't need the `asFun` in the following statement.

File: Mathlib/RingTheory/Finiteness.lean [hunk 4]
```diff
@@ -265,25 +267,31 @@ theorem fg_pi {ι : Type _} {M : ι → Type _} [Finite ι] [∀ i, AddCommMonoi
...
     intro y hy
-    have : y ∈ map f s := by
+    have : y ∈ @map R R M P _ _ _ _ _ _ (RingHom.id _) _ _ instSLMC f s := by
```
Comment (jcommelin) [inline Mathlib/RingTheory/Finiteness.lean:289]: With `etaExperiment`, we don't need the `@` in the following statement.

File: Mathlib/RingTheory/Finiteness.lean [hunk 5]
```diff
@@ -265,25 +267,31 @@ theorem fg_pi {ι : Type _} {M : ι → Type _} [Finite ι] [∀ i, AddCommMonoi
...
       rw [← ht1]
       exact subset_span hy
     rcases mem_map.1 this with ⟨x, hx1, hx2⟩
     exact ⟨x, hx1, hx2⟩
-  have : ∃ g : P → M, ∀ y ∈ t1, g y ∈ s ∧ f (g y) = y :=
+  have : ∃ g : P → M, ∀ y ∈ t1, g y ∈ s ∧ asFun f (g y) = y :=
```
Comment (jcommelin) [inline Mathlib/RingTheory/Finiteness.lean:294]: With `etaExperiment`, we don't need the `asFun` in the following statement.

File: Mathlib/RingTheory/Finiteness.lean [hunk 6]
```diff
@@ -306,7 +314,7 @@ theorem fg_of_fg_map_of_fg_inf_ker {R M P : Type _} [Ring R] [AddCommGroup M] [M
       rw [ht2] at this
       exact this.1
   intro x hx
-  have : f x ∈ map f s := by
+  have : asFun f x ∈ @map R R M P _ _ _ _ _ _ (RingHom.id _) _ _ instSLMC f s := by
```
Comment (jcommelin) [inline Mathlib/RingTheory/Finiteness.lean:317]: With `etaExperiment`, we don't need the `asFun` and `@` in the following statement.

File: Mathlib/RingTheory/Finiteness.lean [hunk 7]
```diff
@@ -354,16 +364,22 @@ theorem fg_induction (R M : Type _) [Semiring R] [AddCommMonoid M] [Module R M]
...
 #align submodule.fg_induction Submodule.fg_induction
 
-set_option synthInstance.etaExperiment true in
 /-- The kernel of the composition of two linear maps is finitely generated if both kernels are and
 the first morphism is surjective. -/
 theorem fg_ker_comp {R M N P : Type _} [Ring R] [AddCommGroup M] [Module R M] [AddCommGroup N]
-    [Module R N] [AddCommGroup P] [Module R P] (f : M →ₗ[R] N) (g : N →ₗ[R] P) (hf1 : f.ker.Fg)
-    (hf2 : g.ker.Fg) (hsur : Function.Surjective f) : (g.comp f).ker.Fg := by
+    [Module R N] [AddCommGroup P] [Module R P] (f : M →ₗ[R] N) (g : N →ₗ[R] P) 
+    -- Porting note: more help both unifying and finding instSLMC
```
Comment (jcommelin) [inline Mathlib/RingTheory/Finiteness.lean:371]: With `etaExperiment`, we don't need the `@`s in the following statement.

File: Mathlib/RingTheory/Finiteness.lean [hunk 8]
```diff
@@ -354,16 +364,22 @@ theorem fg_induction (R M : Type _) [Semiring R] [AddCommMonoid M] [Module R M]
...
+    (hf1 : (@LinearMap.ker R R M N _ _ _ _ _ _ (RingHom.id _) _ instSLMC f).Fg)
+    (hf2 : (@LinearMap.ker R R N P _ _ _ _ _ _ (RingHom.id _) _ instSLMC g).Fg) 
+    (hsur : Function.Surjective <| asFun f) : 
+    (@LinearMap.comp R R R M N P _ _ _ _ _ _ _ _ _ 
+      (RingHom.id _) (RingHom.id _) (RingHom.id _) _ g f).ker.Fg := by
   rw [LinearMap.ker_comp]
   apply fg_of_fg_map_of_fg_inf_ker f
   · rwa [Submodule.map_comap_eq, LinearMap.range_eq_top.2 hsur, top_inf_eq]
-  · rwa [inf_of_le_right (show LinearMap.ker f ≤ comap f (LinearMap.ker g) from comap_mono bot_le)]
+  · rwa [inf_of_le_right (show (@LinearMap.ker R R M N _ _ _ _ _ _ (RingHom.id _) _ instSLMC f) ≤
```
Comment (jcommelin) [inline Mathlib/RingTheory/Finiteness.lean:380]: With `etaExperiment`, we don't need the `@`s in the following statement.

File: Mathlib/RingTheory/Finiteness.lean (hunk 1 above)
Comment (jcommelin) [reply Mathlib/RingTheory/Finiteness.lean:233]:
```suggestion
-- Porting note: With etaExperiment, we don't need the @s in the following statement.
theorem fg_of_fg_map {R M P : Type _} [Ring R] [AddCommGroup M] [Module R M] [AddCommGroup P]
```

Comment (jcommelin) [reply Mathlib/RingTheory/Finiteness.lean:233]:
```suggestion
-- Porting note: With etaExperiment, we don't need the @s in the following statement.
theorem fg_of_fg_map {R M P : Type _} [Ring R] [AddCommGroup M] [Module R M] [AddCommGroup P]
```

File: Mathlib/RingTheory/Finiteness.lean (hunk 2 above)
Comment (jcommelin) [reply Mathlib/RingTheory/Finiteness.lean:279]:
```suggestion
    -- Porting note: Lean having trouble both unifying for SLMC and then synthesizing once unified
    -- With etaExperiment, we don't need the @s in the following statement.
```

File: Mathlib/RingTheory/Finiteness.lean (hunk 3 above)
Comment (jcommelin) [reply Mathlib/RingTheory/Finiteness.lean:287]:
```suggestion
  -- Porting note: With etaExperiment, we don't need the asFun in the following statement.
  have : ∀ y ∈ t1, ∃ x ∈ s, asFun f x = y := by
```

File: Mathlib/RingTheory/Finiteness.lean (hunk 4 above)
Comment (jcommelin) [reply Mathlib/RingTheory/Finiteness.lean:289]:
```suggestion
    -- Porting note: With etaExperiment, we don't need the @ in the following statement.
    have : y ∈ @map R R M P _ _ _ _ _ _ (RingHom.id _) _ _ instSLMC f s := by
```

File: Mathlib/RingTheory/Finiteness.lean (hunk 5 above)
Comment (jcommelin) [reply Mathlib/RingTheory/Finiteness.lean:294]:
```suggestion
  -- Porting note: With etaExperiment, we don't need the asFun in the following statement.
  have : ∃ g : P → M, ∀ y ∈ t1, g y ∈ s ∧ asFun f (g y) = y :=
```

File: Mathlib/RingTheory/Finiteness.lean (hunk 6 above)
Comment (jcommelin) [reply Mathlib/RingTheory/Finiteness.lean:317]:
```suggestion
  -- Porting note: With etaExperiment, we don't need the asFun and @ in the following statement.
  have : asFun f x ∈ @map R R M P _ _ _ _ _ _ (RingHom.id _) _ _ instSLMC f s := by
```

File: Mathlib/RingTheory/Finiteness.lean (hunk 7 above)
Comment (jcommelin) [reply Mathlib/RingTheory/Finiteness.lean:371]:
```suggestion
    -- Porting note: more help both unifying and finding instSLMC
    -- With etaExperiment, we don't need the @s in the following statement.
```

File: Mathlib/RingTheory/Finiteness.lean (hunk 8 above)
Comment (jcommelin) [reply Mathlib/RingTheory/Finiteness.lean:387]:
```suggestion
  -- Porting note With etaExperiment, we don't need the @s in the following `show`.
  · rwa [inf_of_le_right (show (@LinearMap.ker R R M N _ _ _ _ _ _ (RingHom.id _) _ instSLMC f) ≤
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**